src/librustc_codegen_ssa/mir/mod.rs

```diff
@@ -190,6 +191,18 @@ pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
 
     fx.per_local_var_debug_info = fx.compute_per_local_var_debug_info();
 
+    for const_ in &mir.uneval_consts {
```

Comment:
Assuming this PR is addressing https://github.com/rust-lang/rust/issues/67191, it seems like the way this PR is setup, then `cargo check` (`--emit metadata`) would not trigger an error here but it should. In other words, erroneous live constants should be caught in *analysis* before optimizations.

Reply:
yea, we need to go through this list in the collector, too, just like we're doing with in-MIR constants right now. This may get solved automatically by extending the mir visitors to go through the constants in the list via `visit_const`

Reply:
Added this here https://github.com/rust-lang/rust/pull/70820/commits/2f8c3b5c9ed83f704505804bb6f44a810184cd08

Reply:
@oli-obk I don't follow how this was resolved. The code was added to monomorphization logic, but presumably that never runs if we don't reach codegen?

Reply:
I don't see how `cargo check` can trigger an error on these *without* monomorphization. If you look at the test that we're doing all these changes for (https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/assoc_const_generic_impl.rs) you see that without monomorphization we can't possibly report anything. This is inherently a monomorphization time thing.

Reply:
this applies to all other lints reported during mir optimizations, too. Const prop lints for example aren't triggered reliably from `cargo check`. Only if the `optimized_mir` query is called for a function, will const prop lints actually get reported. We could add `optimized_mir` runs for all functions to the analysis query though. In that case we'd also do a best effort evaluation of all constants

Reply:
Note that I feel that running `optimized_mir` on all functions for `analysis` is an orthogonal discussion to this PR and we should probably move it to an issue if you think this is a useful thing to do at all (it will make `cargo check` take more time as optimizations are run)

Reply:
Well this is all very unfortunate... I had neglected associated constants. With them in mind, the options as I see them are:

1. Error differently for polymorphic and monomorphic constants, doing the former post-monomorphization, and the latter pre-monomorphization (inconsistent).

2. Error for both post-monomorphization. (I consider all post-monomorphization errors to be language design bugs.)

3. Error for neither and insert a trap in codegen, but @RalfJung [would probably be even more unhappy about that](https://github.com/rust-lang/rust/pull/67134#discussion_r355697861).

4. Only insert a trap for polymorphic constants (inconsistent).

Reply:
We are doing 1. right now. Erroring post monomorphization for associated constants is what we do to keep our APIs sane. Otherwise we'd need integer range bounds and more just to be able to do simple addition of associated constants inside other associated constants.

I don't like 3 and 4 and am not even sure that 2. is implementable :rofl: 

What we could do at some point in the future is to create a lint that tells you about sites where post monomorphization errors could occur depending on the generic parameters that were given, but that would require symbolic execution of constants and some sort of solver (not necessarily SAT) for them. While that is something we need for more advanced const generics trickeries anyway, I think we can just stay with option 1 until we address post monomorphization errors in general.

Reply:
> We are doing 1. right now. Erroring post monomorphization for associated constants is what we do to keep our APIs sane. Otherwise we'd need integer range bounds and more just to be able to do simple addition of associated constants inside other associated constants.

Yea, this is going to be ungreat for the language spec and inconsistent, but of the bad choices I suppose it's the least bad (😭), so do your thing. :)